### PR TITLE
🐞 fix: fix rt_thread_delay, assert, deadlock

### PR DESCRIPTION
**Pull request by xqyjlj**
## 拉取/合并请求描述：(PR description)

[

#### 为什么提交这份PR (why to submit this PR)

使用rt_thread_suspend_with_flag将自己挂起后，这时候中断是可以响应的，如果在中断中切走，那么就永远不会回来了,

必须先判断是否关闭调度器，再拿自旋锁，否则会出现死锁

自旋锁内禁止使用free,  malloc等内嵌阻塞ipc接口

#### 请提供验证的bsp和config (provide the config and bsp) 

- BSP:

- .config:

- action:

]

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [ ] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [ ] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [ ] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [ ] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [ ] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [ ] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [ ] 代码是高质量的 Code in this PR is of high quality
- [ ] 已经使用[formatting](https://github.com/mysterywolf/formatting) 等源码格式化工具确保格式符合[RT-Thread代码规范](https://github.com/RT-Thread/rt-thread/blob/master/documentation/contribution_guide/coding_style_cn.md) This PR complies with [RT-Thread code specification](https://github.com/RT-Thread/rt-thread/blob/master/documentation/contribution_guide/coding_style_en.md)

### CODE REVIEW

**Comment by xqyjlj** [comment]
此pr需要先合并 #8360

File: src/thread.c [hunk 1]
```diff
@@ -709,7 +733,7 @@ rt_err_t rt_thread_delay_until(rt_tick_t *tick, rt_tick_t inc_tick)
     RT_ASSERT(rt_object_get_type((rt_object_t)thread) == RT_Object_Class_Thread);
 
     /* disable interrupt */
-    level = rt_spin_lock_irqsave(&(thread->spinlock));
+    level = rt_hw_local_irq_disable();
 
     /* reset thread error */
     thread->error = RT_EOK;
```
Comment (geniusgogo) [inline src/thread.c:739]: 这上面如果不是用线程锁，貌似锁不住

Comment (xqyjlj) [reply src/thread.c:739]: 线程锁不锁error，这里只是需要保证 cur_tick 及时获取就好了

Comment (xqyjlj) [reply src/thread.c:736]: 已做修改

File: src/ipc.c [hunk 2]
```diff
@@ -582,7 +586,7 @@ static rt_err_t _rt_sem_take(rt_sem_t sem, rt_int32_t timeout, int suspend_flag)
 
             if (thread->error != RT_EOK)
             {
-                return thread->error;
+                return thread->error > 0 ? -thread->error : thread->error;
```
Comment (geniusgogo) [inline src/ipc.c:589]: 后续像这种地方都应该统一修掉，https://github.com/RT-Thread/rt-thread/pull/8367#discussion_r1434756234

**Review by geniusgogo — APPROVED** [review]

**Pull request merged**